core/src/main/java/io/temporal/samples/autoheartbeat/interceptor/AutoHeartbeatActivityInboundCallsInterceptor.java

```diff
@@ -0,0 +1,106 @@
+import io.temporal.samples.autoheartbeat.AutoHeartbeatUtil;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+
+public class AutoHeartbeatActivityInboundCallsInterceptor
+    extends ActivityInboundCallsInterceptorBase {
+  private ActivityExecutionContext activityExecutionContext;
+  private Duration activityHeartbeatTimeout;
+  private AutoHeartbeatUtil autoHeartbeater;
+  //  private CompletableFuture autoHeartbeatFuture;
+  private ScheduledFuture scheduledFuture;
+
+  //  private ScheduledFuture scheduledFuture;
+
+  public AutoHeartbeatActivityInboundCallsInterceptor(ActivityInboundCallsInterceptor next) {
+    super(next);
+  }
+
+  @Override
+  public void init(ActivityExecutionContext context) {
+    this.activityExecutionContext = context;
+    activityHeartbeatTimeout = activityExecutionContext.getInfo().getHeartbeatTimeout();
+    super.init(context);
+  }
+
+  @Override
+  @SuppressWarnings({"FutureReturnValueIgnored", "CatchAndPrintStackTrace"})
+  public ActivityOutput execute(ActivityInput input) {
+    // If activity has heartbeat timeout defined we want to apply auto-heartbeter
+    if (activityHeartbeatTimeout != null && activityHeartbeatTimeout.getSeconds() > 0) {
+      System.out.println(
+          "Auto heartbeating applied for activity: "
+              + activityExecutionContext.getInfo().getActivityType());
+      autoHeartbeater =
+          new AutoHeartbeatUtil(2, 0, TimeUnit.SECONDS, activityExecutionContext, input);
+      scheduledFuture = autoHeartbeater.start();
+    } else {
+      System.out.println(
+          "Auto heartbeating not being applied for activity: "
+              + activityExecutionContext.getInfo().getActivityType());
+    }
+
+    if (scheduledFuture != null) {
+      CompletableFuture activityExecFuture =
+          CompletableFuture.supplyAsync(() -> super.execute(input));
+      CompletableFuture autoHeartbeatFuture =
+          CompletableFuture.supplyAsync(
+              () -> {
+                try {
+                  return scheduledFuture.get();
+                } catch (Exception e) {
+                  throw new ActivityCanceledException(activityExecutionContext.getInfo());
+                }
+              });
+      try {
+        CompletableFuture.anyOf(autoHeartbeatFuture, activityExecFuture).get();
+      } catch (Exception e) {
+        if (e instanceof ExecutionException) {
+          ExecutionException ee = (ExecutionException) e;
+          if (ee.getCause() instanceof ActivityCanceledException) {
+            throw new ActivityCanceledException(activityExecutionContext.getInfo());
+          }
+        }
+        throw Activity.wrap(e);
+      } finally {
+        if (autoHeartbeater != null) {
+          autoHeartbeater.stop();
+        }
+      }
+    }
+    return super.execute(input);
```

Comment:
I think the activity function is executed twice when there is a heartbeat, here and in line [74](https://github.com/temporalio/samples-java/pull/745/files#diff-7cfc22d0df1ade955f3ba4a93625962f3ae729c79982d16cc7e7e27e71ff4338R74) of this same file